notes.md

```diff
@@ -0,0 +1,124 @@
+* There is an external script that requested the minting be done for some address, and awaits the tokenIds from the `Minted` event, which it records against the user in a list of owned certificate numbers.
+
+c) `Token Buys Share immediately` && `One token = one certificate has many shares` && `Certificate numbers come from the blockchain`
+
+Same as (a) but with the following exceptions.
+
+* extend (a) with a base mapping of `shareNumbers` to `tokenIds`, and mapping of `tokenIds` to an array of `shareNumbers`.
+* When a `Token` is minted for an address, the `tokenId` is provided by the CrowdsaleContract but the share number range needs to be provided as an encoded bytes array looking like `[1,2,53, 1000..5000, 6750]`. The CrowdsaleContract needs to turn this into individual `shareIds`.
+* The `tokenIds` are used as the certificate numbers for the associated share certificates which include share ranges.
+* It is the `CrowdsaleContract`'s job to create the `tokenId`, representing the certificate number associated with it and pass that information into each `Token` when it is `mint`ed. It's the job of the script requesting the minting process to provide the share ranges and listen for the appropriate `Minted` event coming from the CrowdsaleContract carrying back that same `tokenId` for checking purposes, then recording the transaction receipt with the users' other data in the offchain-database.
+* Each `Token` maps to a certificate with a number, that in turn describes a share range. The Tokens need to be issued one at a time by the Crowdsale visa the `mint` function.
+* There is an external script that requested the minting be done for some address, and awaits the tokenIds from the `Minted` event, which it records against the user in a list of owned certificate numbers.
+* In this case the owner could sell individual shares somehow, say offchain, and the range would need to be updated from offchain, in which case the Token would need a `transferShare(uint shareId) public onlyShareholder` function and an `onlyShareholder` modifier.
+
+d) `Token Buys Share immediately` && `One token = one certificate has many shares` && `Certificate numbers are generated externally`
+
+Same as (c) but the external script injects both the share range and handle the certificate numbers like in (b)
+
+e) `Token Buys rights to a share, when the crowdsale is finished.` && `One token = one certificate = one share` && `Certificate numbers come from the blockchain`
+
+You'd do this if you want to ensure a minimum cap is reached before issuing stock.
+
+* same as (a) but don't actually do the `Transfers` in the `mint` function until the `finalise()` function is called.
+
+f) `Token Buys rights to a share, when the crowdsale is finished.` && `One token = one certificate = one share` && `Certificate numbers are generated externally`
```

Comment:
I think I like f the most so far, but whatever gives the most flexible implementation.  Kind of unrelated, is it the case that we have to assume the number of shares can change in the future or we just stick with say 150k shares? It seems much easier to work with constant shares.

Reply:
@tavakyan ...unfortunately the number of shares can change in the future.

There should be a function for "stock splitting"... but couldn't one theoretically just launch a new ledger with different params? The good news is that even if the distributed stock ledger breaks at some corner solution, investors are still protected off-chain.

In conversation with Jeff tomorrow, 'f' may be most practical but i think i prefer 'b' at the moment...